packages/fiber/src/core/reconciler.tsx

```diff
@@ -62,7 +64,23 @@ interface HostConfig {
 }
 
 const catalogue: Catalogue = {}
-export const extend = (objects: Partial<Catalogue>): void => void Object.assign(catalogue, objects)
+
+let i = 0
+
+export const extend = <T extends Catalogue | ConstructorRepresentation>(
+  objects: T,
+): T extends ConstructorRepresentation ? React.ExoticComponent<ThreeElement<T>> : void => {
+  if (typeof objects === 'function') {
+    const Component = `${i++}`
+    catalogue[Component] = objects
+
+    // Returns a component whose name will be inferred in devtools
+    // @ts-ignore
+    return React.forwardRef({ [objects.name]: (props, ref) => <Component {...props} ref={ref} /> }[objects.name])
+  } else {
+    return void Object.assign(catalogue, objects) as any
+  }
+}
```

Comment:
This is a little bit of a hack, but will ensure that `extend(...)` won't face collisions since it's not valid for identifiers to start with a non-alphanumeric character in JS. `{ [key]: value }[key]` will also create a named function inline so it will appear correctly in devtools rather than just a number.